File: 01-Programming-Languages/R/Data-Structures/Stack/stack.R
Stack <- setRefClass("Stack",
  fields = list(
    items = "list"
  ),
  methods = list(
    initialize = function() {
      .self$items <- list()
    },
    
    push = function(item) {
      .self$items <- c(.self$items, list(item))
    },
    
    pop = function() {
      if (length(.self$items) == 0) return(NULL)
      item <- .self$items[[length(.self$items)]]
      .self$items <- .self$items[-length(.self$items)]
      item
    },
    
    peek = function() {
      if (length(.self$items) == 0) return(NULL)
      .self$items[[length(.self$items)]]
    },
    
    is_empty = function() {
      length(.self$items) == 0
    },
    
    size = function() {
      length(.self$items)
    }
  )
)

stack <- Stack$new()
stack$push(1)
stack$push(2)
stack$push(3)

cat("Stack size:", stack$size(), "\n")
cat("Popped:", stack$pop(), "\n")
cat("Top:", stack$peek(), "\n")

File: 01-Programming-Languages/Kotlin/Data-Structures/Stack/stack.kt
class Stack<T> {
    private val items = mutableListOf<T>()

    fun push(item: T) {
        items.add(item)
    }

    fun pop(): T? {
        return if (items.isNotEmpty()) items.removeAt(items.size - 1) else null
    }

    fun peek(): T? {
        return items.lastOrNull()
    }

    fun isEmpty(): Boolean = items.isEmpty()

    fun size(): Int = items.size
}

fun main() {
    val stack = Stack<Int>()
    stack.push(1)
    stack.push(2)
    stack.push(3)

    println("Stack size: ${stack.size()}")
    println("Popped: ${stack.pop()}")
    println("Top: ${stack.peek()}")
}